Source/Core/Duality/Utility/Coroutines/CoroutineManager.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Duality.Utility.Pooling;
+
+namespace Duality
+{
+	public class CoroutineManager
+	{
+		private static QueuedPool<Coroutine> pool = new QueuedPool<Coroutine>();
+		private readonly List<Coroutine> coroutines = new List<Coroutine>();
+		private readonly List<Coroutine> scheduled = new List<Coroutine>();
+		private readonly List<Coroutine> trashcan = new List<Coroutine>();
+		private readonly Dictionary<Coroutine, Exception> lastFrameErrors = new Dictionary<Coroutine, Exception>();
+
+		public IEnumerable<Coroutine> Coroutines
+		{
+			get { return this.coroutines.Concat(this.scheduled); }
+		}
+
+		public IEnumerable<KeyValuePair<Coroutine, Exception>> LastFrameErrors
+		{
+			get { return this.lastFrameErrors; }
+		}
+
+		public Coroutine StartNew(IEnumerator<IWaitCondition> enumerator, string name)
+		{
+			Coroutine coroutine = pool.GetOne();
+			coroutine.Setup(enumerator, name);
+
+			this.scheduled.Add(coroutine);
+			return coroutine;
+		}
+
+		public bool IsCoroutineListed(Coroutine c)
+		{
+			return this.Coroutines.Contains(c);
+		}
+
+		public void Clear()
+		{
+			foreach (Coroutine c in this.coroutines)
+				c.Cancel();
+
+			this.scheduled.Clear();
+			this.coroutines.Clear();
+		}
+
+		public void Update()
+		{
+			this.lastFrameErrors.Clear();
+
+			this.coroutines.AddRange(this.scheduled);
+			this.scheduled.Clear();
+
+			foreach (Coroutine c in this.coroutines)
+			{
+				c.Update();
+
+				if (c.Status == CoroutineStatus.Error)
+					this.lastFrameErrors.Add(c, c.LastException);
+				if (c.Status != CoroutineStatus.Running && c.Status != CoroutineStatus.Paused)
+					this.trashcan.Add(c);
+			}
+
+			foreach (Coroutine c in this.trashcan)
+			{
+				this.coroutines.Remove(c);
+				pool.ReturnOne(c);
+			}
```

Comment:
Couple ideas that I think would make this section more performant:

1. Currently, removing each completed coroutine from `trashcan` is O(n) to find the coroutine to delete and then O(n) to actually remove it from the list. If `trashcan` was `List<int>` storing the indexes of completed coroutines, you could iterate through `trashcan` in reverse and remove from `this.coroutines` by index (O(1) to lookup, O(n) to remove).
2. You could make the actual erasing of the coroutine from `this.coroutines` faster if it was safe to reorder coroutines by swapping completed coroutines with the last incomplete coroutine and then truncating `this.coroutines` at the end.
3. You could even go as far as to remove the use of a separate pool list and simply store the pool of coroutines in `this.coroutines` - keeping completed coroutines at the end and reusing them when necessary. Then when a coroutine completes all you would have to do is swap the completed coroutine with the last incomplete one.

2 & 3 may cause confusion if anyone using coroutines expects them to update in the same order every frame. That's a dangerous assumption for someone to make, but it could happen. It may be worth optimizing though as I remember with Unity's coroutines people were bitten by their performance quite often - we should avoid that here.

Reply:
Why do you say that removing is O(n) + O(n)? the trashcan contains the same Coroutine that has to be removed from the list, no search is being done 🤔

Reply:
When you use Remove(instance) on a list it first has to search where that instance is in the list hence the O(n).

Reply:
No no, I understand where the O(n) is coming from in the Remove call, I was just wondering where the other O(n) is.. the only other loop is during the update part, but that can't be avoided.

I'll go ahead and do a removal by index for now - I'll try with a specialized pool to try and implement version 3. I think it's just safe to explicitly state that execution order of Coroutines is not guaranteed

Reply:
Changed to Queues to have O(1) all around the board